File: test/Words.Test/LetterBoxStrSearchTest.cs
// <copyright file="LetterBoxStrSearchTest.cs" company="Brian Rogers">
// Copyright (c) Brian Rogers. All rights reserved.
// </copyright>

namespace Words.Test
{
    using System.Collections.Generic;
    using FluentAssertions;
    using Xunit;

    public sealed class LetterBoxStrSearchTest
    {
        [Fact]
        public void EmptyTrieFindsNothing()
        {
            StrTrie trie = new StrTrie();
            LetterBoxStrSearch search = New(trie);

            FindWords(search).Should().BeEmpty();
        }

        [Fact]
        public void OneValueTrieFindsOneWord()
        {
            StrTrie trie = new StrTrie();
            trie.Add(Str.Parse("ALE"));
            LetterBoxStrSearch search = New(trie);
            List<string> found = new List<string>();

            FindWords(search).Should().BeEquivalentTo("ALE:100010000001");
        }

        [Fact]
        public void TwelveValueTrieFindsAllWords()
        {
            StrTrie trie = new StrTrie();
            trie.Add(Str.Parse("ALE"));
            trie.Add(Str.Parse("BEG"));
            trie.Add(Str.Parse("CEL"));
            trie.Add(Str.Parse("DAH"));
            trie.Add(Str.Parse("ELF"));
            trie.Add(Str.Parse("FIB"));
            trie.Add(Str.Parse("GAL"));
            trie.Add(Str.Parse("HAD"));
            trie.Add(Str.Parse("ICE"));
            trie.Add(Str.Parse("JIB"));
            trie.Add(Str.Parse("KAE"));
            trie.Add(Str.Parse("LIE"));
            LetterBoxStrSearch search = New(trie);
            List<string> found = new List<string>();

            FindWords(search).Should().BeEquivalentTo(
                "ALE:100010000001",
                "BEG:010010100000",
                "CEL:001010000001",
                "DAH:100100010000",
                "ELF:000011000001",
                "FIB:010001001000",
                "GAL:100000100001",
                "HAD:100100010000",
                "ICE:001010001000",
                "JIB:010000001100",
                "KAE:100010000010",
                "LIE:000010001001");
        }

        [Fact]
        public void SearchDoesNotReturnInvalidMoves()
        {
            StrTrie trie = new StrTrie();
            trie.Add(Str.Parse("ABC"));
            trie.Add(Str.Parse("DEF"));
            trie.Add(Str.Parse("GHI"));
            trie.Add(Str.Parse("JKL"));
            trie.Add(Str.Parse("MOW"));
            trie.Add(Str.Parse("ALA"));
            LetterBoxStrSearch search = New(trie);
            List<string> found = new List<string>();

            FindWords(search).Should().BeEquivalentTo("ALA:100000000001");
        }

        [Fact]
        public void SearchDoesNotAttemptToBuildTooLongWords()
        {
            StrTrie trie = new StrTrie();
            trie.Add(Str.Parse("LA"));
            trie.Add(Str.Parse("LALA"));
            trie.Add(Str.Parse("LALALA"));
            trie.Add(Str.Parse("LALALALA"));
            trie.Add(Str.Parse("LALALALALA"));
            trie.Add(Str.Parse("LALALALALALA"));
            LetterBoxStrSearch search = New(trie);
            List<string> found = new List<string>();

            FindWords(search).Should().BeEquivalentTo(
                "LA:100000000001",
                "LALA:100000000001",
                "LALALA:100000000001",
                "LALALALA:100000000001",
                "LALALALALA:100000000001",
                "LALALALALALA:100000000001");
        }

        private static LetterBoxStrSearch New(StrTrie trie)
        {
            return new LetterBoxStrSearch(trie, new LetterBoxStr(Str.Parse("ABCDEFGHIJKL")));
        }

        private static IList<string> FindWords(LetterBoxStrSearch search)
        {
            List<string> found = new List<string>();

            search.Run((w, v) => found.Add(w + ":" + v));

            return found;
        }
    }
}

File: test/Words.Native.Test/LetterBoxStrSearchTest.cpp
#include <gtest/gtest.h>
#include "LetterBoxStrSearch.h"
#include <vector>
#include <bitset>
#include <string>

using namespace std;

namespace Words
{
    LetterBoxStrSearch Init(const StrTrie& trie)
    {
        return LetterBoxStrSearch(trie, LetterBoxStr("ABCDEFGHIJKL"));
    }

    void FindWords(const LetterBoxStrSearch& search, initializer_list<string> expected)
    {
        vector<string> found;
        vector<string> expectedFound(expected);

        search.run([&found](Str w, Vertices v)
            {
                stringstream ss;
                ss << w << ":" << v;
                found.push_back(ss.str());
            });

        ASSERT_EQ(expected.size(), found.size());
        for (int i = 0; i < expected.size(); ++i)
        {
            ASSERT_STREQ(expectedFound[i].c_str(), found[i].c_str());
        }
    }

    TEST(LetterBoxStrSearchTest, EmptyTrieFindsNothing)
    {
        StrTrie trie;
        LetterBoxStrSearch search(Init(trie));

        FindWords(search, {});
    }

    TEST(LetterBoxStrSearchTest, OneValueTrieFindsOneWord)
    {
        StrTrie trie;
        trie.insert("ALE");
        LetterBoxStrSearch search(Init(trie));

        FindWords(search, { "ALE:100000010001" });
    }

    TEST(LetterBoxStrSearchTest, TwelveValueTrieFindsAllWords)
    {
        StrTrie trie;
        trie.insert("ALE");
        trie.insert("BEG");
        trie.insert("CEL");
        trie.insert("DAH");
        trie.insert("ELF");
        trie.insert("FIB");
        trie.insert("GAL");
        trie.insert("HAD");
        trie.insert("ICE");
        trie.insert("JIB");
        trie.insert("KAE");
        trie.insert("LIE");
        LetterBoxStrSearch search(Init(trie));

        FindWords(
            search,
            {
                "ALE:100000010001",
                "BEG:000001010010",
                "CEL:100000010100",
                "DAH:000010001001",
                "ELF:100000110000",
                "FIB:000100100010",
                "GAL:100001000001",
                "HAD:000010001001",
                "ICE:000100010100",
                "JIB:001100000010",
                "KAE:010000010001",
                "LIE:100100010000"
            });
    }

    TEST(LetterBoxStrSearchTest, SearchDoesNotReturnInvalidMoves)
    {
        StrTrie trie;
        trie.insert("ABC");
        trie.insert("DEF");
        trie.insert("GHI");
        trie.insert("JKL");
        trie.insert("MOW");
        trie.insert("ALA");
        LetterBoxStrSearch search(Init(trie));

        FindWords(search, { "ALA:100000000001" });
    }

    TEST(LetterBoxStrSearchTest, SearchDoesNotAttemptToBuildTooLongWords)
    {
        StrTrie trie;
        trie.insert("LA");
        trie.insert("LALA");
        trie.insert("LALALA");
        trie.insert("LALALALA");
        trie.insert("LALALALALA");
        trie.insert("LALALALALALA");
        LetterBoxStrSearch search(Init(trie));

        FindWords(
            search,
            {
                "LA:100000000001",
                "LALA:100000000001",
                "LALALA:100000000001",
                "LALALALA:100000000001",
                "LALALALALA:100000000001",
                "LALALALALALA:100000000001"
            });
    }
}
